### PR TITLE
Add ERC: Tamperproof Web Immutable Transaction (TWIT)

### PR DESCRIPTION
**Pull request by uni-guillaume**
This EIP written in close collaboration with @remarks propose a new security mechanism to guarantee data integrity between dapps and wallets.
It introduces a new RPC method to be implemented by wallets, `wallet_signedRequest`, that
enables dapps to interact with wallets in a tamperproof manner via "signed requests". The
dapp associates a public key with its DNS record and uses the corresponding private key to
sign payloads sent to the wallet via `wallet_signedRequest`. Wallets can then use use the
public key in the DNS record to validate the integrity of the payload.
The wallet can subsequently confirm data integrity to their users

### CODE REVIEW

File: ERCS/erc-XXXX.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
+---
+eip: XXXX
```
Comment (abcoathup) [inline ERCS/erc-XXXX.md:2]:
```suggestion
eip: 7754
```
Assigning next sequential EIP/ERC/RIP number.

Please also update the filename.

File: ERCS/erc-XXXX.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+title: Tamperproof Web Immutable Transaction (TWIT)
+description: Provides a mechanism for DAP to use the API defined in EIP-1193 in a tamperproof way
+author: Erik Marks, Guillaume Grosbois
```
Comment (abcoathup) [inline ERCS/erc-XXXX.md:5]:
```suggestion
author: Erik Marks, Guillaume Grosbois (@uni-guillaume)
```

File: ERCS/erc-XXXX.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+discussions-to: TODO
```
Comment (abcoathup) [inline ERCS/erc-XXXX.md:6]: Please create a topic in Eth Magicians.  https://ethereum-magicians.org/
It can just contain a link to this PR.

Comment (uni-guillaume) [reply ERCS/erc-XXXX.md]: will do.

**Review by petvaizAkhtar — APPROVED** [review]

**Comment by uni-guillaume** [comment]
@abcoathup @petvaizAkhtar  : is the PR good enough to be merged ?

**Comment by uni-guillaume** [comment]
@axic , @g11tech, @SamWilsn, @xinbenlv: do you have any feedback ?

File: ERCS/erc-7754.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Overview
+
+We propose to use the dapp's domain certificate of a root of trust to establish a trust chain as follow:
+
+1. The user's browser verifies the domain certificate and displays appropriate warnings if overtaken
+2. The DNS record of the dapp hosts a TXT field pointing to a URL where a JSON manifest is hosted
+   - This file SHOULD be at a well known address such as `https://domain.com/.well-known/twit.json`
```
Comment (SamWilsn) [inline ERCS/erc-7754.md:57]:
```suggestion
   - This file SHOULD be at a well known address such as `https://example.com/.well-known/twit.json`
```

File: ERCS/erc-7754.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+5. The original payload, signature, and public key id are sent to the wallet via the `wallet_signedRequest` RPC method
+6. The wallet verifies the signature before processing the request normally
+
+### Wallet integration
+
+#### Key discovery
+
+Attested public keys are necessary for the chain of trust to be established.
+Since this is traditionally done via DNS certificates, we propose the addition of a DNS record containing the public keys.
+This is similar to DKIM, but the use of a manifest file provides more flexibility for future improvements, as well as support for multiple algorithm and key pairs.
```
Comment (SamWilsn) [inline ERCS/erc-7754.md:69]: Readers might not be familiar with DKIM. You could do something like:

```suggestion
This is similar to [RFC 6376](https://www.rfc-editor.org/rfc/rfc6376.html)'s DKIM, but the use of a manifest file provides more flexibility for future improvements, as well as support for multiple algorithm and key pairs.
```

File: ERCS/erc-7754.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+
+Similarly to standard JWT practices, the wallet could eagerly cache dapp keys.
+However, in the absence of a revocation mechanism, a compromised key could still be used until caches have expired.
+To mitigate this, wallets SHOULD NOT cache dapp public keys for more than 2 hours.
+This practice establishes a relatively short vulnerability window, and manageable overhead for both wallet and dapp maintainers.
+
+Example DNS record for `my-crypto-dapp.org`:
```
Comment (SamWilsn) [inline ERCS/erc-7754.md:76]:
```suggestion
Example DNS record for `my-crypto-dapp.invalid`:
```

I'd suggest using reserved domain names (like `example.com`, `foo.invalid`, etc.) for your examples.

File: ERCS/erc-7754.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+
+### Wallet integration
+
+#### Key discovery
+
+Attested public keys are necessary for the chain of trust to be established.
+Since this is traditionally done via DNS certificates, we propose the addition of a DNS record containing the public keys.
+This is similar to DKIM, but the use of a manifest file provides more flexibility for future improvements, as well as support for multiple algorithm and key pairs.
+
+Similarly to standard JWT practices, the wallet could eagerly cache dapp keys.
```
Comment (SamWilsn) [inline ERCS/erc-7754.md:71]: Similar comment here about JWT. Either expand the abbreviation or link to a definition.

File: ERCS/erc-7754.md [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+#### RPC method
+
+The parameters of `wallet_signedRequest` are specified by this TypeScript interface:
+
+```typescript
+type RequestPayload<Params> = { method: string; params: Params };
+
+type SignedRequestParameters<Params> = [
+  requestPayload: RequestPayload<Params>,
+  signature: `0x${string}`, // TODO: Is this the format we want?
```
Comment (SamWilsn) [inline ERCS/erc-7754.md:130]: If you use HTML-style comments (outside the code fence), the linter will make sure you deal with them before moving into Review (eg. `<!-- TODO -->`.)

File: ERCS/erc-7754.md [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+4. If TWIT is configured and the `wallet_signedRequest` method is used with the parameters `requestPayload`, `signature` and `keyId` then:
+   a. The wallet MAY display a visible cue indicating that this interaction is signed
+   b. The wallet MUST verify that the keyId exists in the TWIT manifest and find the associated key record
+   c. From the key record, the wallet MUST use the `alg` field and the `publicKey` field to verify `requestPayload` integrity by calling `crypto.verify(alg, key, signature, requestPayload)`
+   d. If the signature is invalid, the wallet MUST display a visible and actionable warning to the user
+   i. If the user opts to ignore the warning, then proceed to call `request` with the argument `requestPayload`
+   ii. If the user opts to cancel, then the wallet MUST cancel the call
+   e. If the signature is valid, the wallet MUST call `request` with the argument `requestPayload`
+
+#### Example method implementation (wallet)
```
Comment (SamWilsn) [inline ERCS/erc-7754.md:175]: Should probably move this to the Reference Implementation section and link to it there from here.

Comment (SamWilsn) [reply ERCS/erc-7754.md:175]: It's up to you. It's short enough that it's probably fine to stay where it is.

**Review by SamWilsn — APPROVED** [review]

**Review by Magken91 — APPROVED** [review]

**Pull request merged**